### ISSUE TITLE
Bert inference error:INTERNAL ERROR: Assertion failed: findIter != mFunctions.end()

### ISSUE THREAD

**Issue by wellinxu** [NVIDIA/TensorRT#1189]
## Description

We have got the tiny-bert model of tensorrt, and we report an error when inferring.

[TensorRT] INTERNAL ERROR: Assertion failed: findIter != mFunctions.end()
/workspace/projects/red_tensorrt/TensorRT/plugin/bertQKVToContextPlugin/fused_multihead_attention.h:261
Aborting...

Aborted (core dumped)

Because of #1096 ,We recompiled lib，and when useing the v100,everything is ok.

## Environment

**TensorRT Version**: 7.2.2.3
**NVIDIA GPU**: A100
**NVIDIA Driver Version**:455.23.05 
**CUDA Version**: 11.1
**CUDNN Version**: 8
**Operating System**:ubuntu 18.04 
**Python Version (if applicable)**: 3.6.9
**Tensorflow Version (if applicable)**: gpu 2.4
**PyTorch Version (if applicable)**: 
**Baremetal or Container (if so, version)**:

**Comment by wellinxu** [comment]
when using bert-base model of tensorrt,it can be inferred.(Regardless of the accuracy of the results)

config of tiny-bert:
{
  "attention_probs_dropout_prob": 0.1,
  "hidden_act": "gelu",
  "hidden_dropout_prob": 0.1,
  "hidden_size": 312,
  "initializer_range": 0.02,
  "intermediate_size": 1200,
  "layer_norm_eps": 1e-12,
  "max_position_embeddings": 512,
  "num_attention_heads": 12,
  "num_hidden_layers": 4,
  "pooler_fc_size": 312,
  "pooler_num_attention_heads": 12,
  "pooler_num_fc_layers": 3,
  "pooler_size_per_head": 128,
  "pooler_type": "first_token_transform",
  "type_vocab_size": 2,
  "vocab_size": 21128
}

**Comment by ttyio** [comment]
@wellinxu , did you see failure before custom build with the fix?

**Comment by wellinxu** [comment]
@ttyio didn't see any other failure

**Comment by ttyio** [comment]
It is strange, If build the 7.2 branch w/o any changes, could you still see the failure? the commit you mentioned should not related to it. 

also what's the sequence length in your test?

**Comment by wellinxu** [comment]
@ttyio when seq_len = 128 or 384,we will get this error. when seq_len = 256,it can be inferred,but its speed is slow .
a100, seq_len=256, time=7.228ms,fp16
v100,seq_len=256,time=9.142ms,fp16

i will try other branch later,but will the error of #1096 be in other branch?

**Comment by ttyio** [comment]
Hello @wellinxu , 
256 only exist in experiment feature: variable-sequence in https://github.com/NVIDIA/TensorRT/tree/release/7.2/demo/BERT#variable-sequence-length

In fixed length case:
for seq_len = 256, it is running using unfused kernel, you need cherry-pick the fix, but it never go to the failure path in your description. it is for fused kernel.
for seq_len = 128 or 384, it is running fused kernel, no need to cherry-pick the fix.

**Comment by wellinxu** [comment]
@ttyio  sorry,when rebulid after #1096, there is failure.

CMakeFiles/third_party.protobuf.dir/build.make:90: recipe for target 'third_party.protobuf/src/third_party.protobuf-stamp/third_party.protobuf-download' failed
make[2]: *** [third_party.protobuf/src/third_party.protobuf-stamp/third_party.protobuf-download] Error 1
CMakeFiles/Makefile2:136: recipe for target 'CMakeFiles/third_party.protobuf.dir/all' failed
make[1]: *** [CMakeFiles/third_party.protobuf.dir/all] Error 2
Makefile:151: recipe for target 'all' failed
make: *** [all] Error 2

**Comment by wellinxu** [comment]
hello @ttyio 

model:tiny-bert,batch_size=75,seq_len=256,fp16

Using model with seq_len = 256 and fp16 can be correctly inferred on A100, but its speed is relatively slow (slower than onnx runtime), which is only 20% faster than V100 (the speed of onnx runtime on A100 is about twice that of V100) https://github.com/NVIDIA/TensorRT/tree/release/7.2/demo/BERT#inference-performance-nvidia-a100-40gb  shows that the speed of A100 is 100% faster than that of V100. Is this normal?

中文表述：
使用seq_len=256，fp16转的模型可以在A100上正确推断，但是速度比较慢（比onnx runtime还慢），只比V100快20%（onnx runtime在A100上的速度大概是V100的两倍）,而根据https://github.com/NVIDIA/TensorRT/tree/release/7.2/demo/BERT#inference-performance-nvidia-a100-40gb 的结果比较，A100的速度比v100快100%以上。这正常吗？

**Comment by ttyio** [comment]
Hello @wellinxu 
sequence length is important factor for the perf in demobert. this is because:

For multi-attention-head, we provide fused kernel for sequence length = 128, 384 on the fixed length version, https://github.com/NVIDIA/TensorRT/blob/release/7.2/demo/BERT/builder.py
we provide fused kernel for sequence length = 64, 96, 128, 256, 384 on the variable length version, https://github.com/NVIDIA/TensorRT/tree/release/7.2/demo/BERT#variable-sequence-length

Could you give a try on the variable sequence length? thanks

**Comment by ttyio** [comment]
@wellinxu  for the build failure, have you tried build inside the docker? https://github.com/NVIDIA/TensorRT/tree/release/7.2#setting-up-the-build-environment

**Comment by wellinxu** [comment]
hello @ttyio 
1. We have tried  the variable sequence length ,but the result of output is wrong. We will check it again.Is there a more detailed document about variable sequence length?
2. and yes，we build inside the docker, but dockerfile is different.
Here is our dockerfile(based on ubuntu-18.04.Dockerfile):


```

ARG CUDA_VERSION=11.1
ARG OS_VERSION=18.04

FROM nvidia/cuda:${CUDA_VERSION}-cudnn8-devel-ubuntu${OS_VERSION}
LABEL maintainer="NVIDIA CORPORATION"

ENV TRT_VERSION 7.2.2.3
SHELL ["/bin/bash", "-c"]

# Setup user account
#ARG uid=1000
#ARG gid=1000
#RUN groupadd -r -f -g ${gid} trtuser && useradd -r -u ${uid} -g ${gid} -ms /bin/bash trtuser
#RUN usermod -aG sudo trtuser
#RUN echo 'trtuser:nvidia' | chpasswd
#RUN mkdir -p /workspace && chown trtuser /workspace

# Install requried libraries
RUN apt-get update && apt-get install -y software-properties-common
#RUN add-apt-repository ppa:ubuntu-toolchain-r/test
RUN apt-get update && apt-get install -y --no-install-recommends \
    libcurl4-openssl-dev \
    wget \
    zlib1g-dev \
    git \
    pkg-config \
    sudo \
    ssh \
    libssl-dev \
    pbzip2 \
    pv \
    bzip2 \
    unzip \
    devscripts \
    lintian \
    fakeroot \
    dh-make \
    build-essential

# Install python3
RUN apt-get install -y --no-install-recommends \
      python3 \
      python3-pip \
      python3-dev \
      python3-wheel &&\
    cd /usr/local/bin &&\
    ln -s /usr/bin/python3 python &&\
    ln -s /usr/bin/pip3 pip;

# Install TensorRT
RUN cd /tmp &&\
    wget https://developer.download.nvidia.com/compute/machine-learning/repos/ubuntu1804/x86_64/nvidia-machine-learning-repo-ubuntu1804_1.0.0-1_amd64.deb &&\
    dpkg -i nvidia-machine-learning-repo-*.deb && apt-get update
RUN v="${TRT_VERSION%.*}-1+cuda${CUDA_VERSION%.*}" &&\
    apt-get install -y libnvinfer7=${v} libnvinfer-plugin7=${v} libnvparsers7=${v} libnvonnxparsers7=${v} libnvinfer-dev=${v} libnvinfer-plugin-dev=${v} libnvparsers-dev=${v} python3-libnvinfer=${v} &&\
    apt-mark hold libnvinfer7 libnvinfer-plugin7 libnvparsers7 libnvonnxparsers7 libnvinfer-dev libnvinfer-plugin-dev libnvparsers-dev python3-libnvinfer

# Install PyPI packages
RUN pip3 install --upgrade pip
RUN pip3 install setuptools>=41.0.0
COPY requirements.txt /tmp/requirements.txt
RUN pip3 install -i https://pypi.douban.com/simple/  -r /tmp/requirements.txt  

# Install Cmake
COPY cmake-3.14.4-Linux-x86_64.sh /tmp/cmake-3.14.4-Linux-x86_64.sh
RUN cd /tmp && \
    chmod +x cmake-3.14.4-Linux-x86_64.sh && \
    ./cmake-3.14.4-Linux-x86_64.sh --prefix=/usr/local --exclude-subdir --skip-license && \
    rm ./cmake-3.14.4-Linux-x86_64.sh

# Download NGC client
COPY ngccli_cat_linux.zip /usr/local/bin/ngccli_cat_linux.zip
RUN cd /usr/local/bin &&  unzip ngccli_cat_linux.zip && chmod u+x ngc && rm ngccli_cat_linux.zip ngc.md5 
RUN echo "no-apikey\nascii\n"  
RUN ngc config set


# Set environment and working directory
ENV TRT_LIBPATH /usr/lib/x86_64-linux-gnu
ENV TRT_OSSPATH /workspace/TensorRT
ENV LD_LIBRARY_PATH="${LD_LIBRARY_PATH}:${TRT_OSSPATH}/build/out:${TRT_LIBPATH}"
WORKDIR /workspace

#USER trtuser
RUN ["/bin/bash"]
```

**Comment by ttyio** [comment]
@wellinxu 
1. sorry currently there is no documentation, you need read the code by yourself.
2. what about use the opensource build environment to build the code, then  copy the binary to your container?

**Comment by wellinxu** [comment]
hello @ttyio 
When using the variable sequence length, the results of bert-base remain unchanged,but the results of tiny-bert are random.

我们发现，使用变长序列的时候，bert-base转换的模型，多次运行结果保持一直，但是tiny-bert转换的模型，每次跑出来的结果都不一致

**Comment by nvpohanh** [comment]
@wellinxu Could you try TRT 8.2/8.4 and see if the issue still exists? If it does, we will debug it. Thanks

**Comment by nvpohanh** [comment]
Closing due to >14 days without activity. Please feel free to reopen if the issue still exists. Thanks
